src/libopensc/pkcs15-eoi.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * Support for the eOI card.
+ *
+ * Copyright (C) 2022 Luka Logar <luka.logar@iname.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <string.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#if defined(ENABLE_SM) && defined(ENABLE_OPENPACE)
+
+#include "log.h"
+#include "pkcs15.h"
+#include "cards.h"
+#include "card-eoi.h"
+
+int sc_pkcs15emu_eoi_init_ex(struct sc_pkcs15_card *p15card, struct sc_aid *aid)
+{
+	struct sc_card *card = p15card->card;
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+	struct sc_pkcs15_search_key sk;
+	struct sc_pkcs15_object *objs[MAX_OBJECTS];
+	int i, j, len;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (card->type != SC_CARD_TYPE_EOI && card->type != SC_CARD_TYPE_EOI_CONTACTLESS)
+		return SC_ERROR_WRONG_CARD;
+
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	/*
+	 * Some of the data is not accessible over the unencrypted channel
+	 * when contactless reader is used. So start SM now.
+	 */
+	if (card->type == SC_CARD_TYPE_EOI_CONTACTLESS) {
+		int r = card->sm_ctx.ops.open(card);
+		if (r != SC_SUCCESS)
+			LOG_FUNC_RETURN(card->ctx, r);
+	}
+
+	/*
+	 * Get the card objects, so we can manipulate them. See below
+	 */
+	LOG_TEST_RET(card->ctx, sc_pkcs15_bind_internal(p15card, aid),
+		"sc_pkcs15_bind_internal failed");
+
+	/*
+	 * PIN objects:
+	 * 1) Find the "Card CAN" PIN and store it's path, so we'll be able to fetch the CAN and do the PACE auth
+	 * 2) Add PIN's auth_info->path to the list of paths that can fail on select. sc_pin_cmd would break otherwise
+	 */
+	memset(&sk, 0, sizeof(sk));
+	sk.class_mask = SC_PKCS15_SEARCH_CLASS_AUTH;
+	len = sc_pkcs15_search_objects(p15card, &sk, (struct sc_pkcs15_object **)&objs, MAX_OBJECTS);
+	for (i = 0, j = 0; i < len; i++) {
+		struct sc_pkcs15_auth_info *auth_info = (struct sc_pkcs15_auth_info *)objs[i]->data;
+		if (auth_info && auth_info->auth_id.len == 8 && !strcmp((char*)auth_info->auth_id.value, "Card CAN")) {
+			auth_info->path.type = SC_PATH_TYPE_PATH;
+			/* Read the file that contains serial and encrypted CAN */
+			if (sc_pkcs15_read_file(p15card, &auth_info->path, &privdata->enc_can.value, &privdata->enc_can.len, 0) == SC_SUCCESS) {
```

Comment:
What's the security model behind making the CAN more or less freely available?

Is it possible that `magic_key` is defined by some proprietary external application and may change in future versions? And if so, should we *store* the CAN in encrypted form in this file to allow parallel use with the "official" PKCS#11 library?

Reply:
Don't ask me. It's only available over the contact interface though.

I am just mimicking what 'official' PKCS#11 library  is doing . I am not 100% sure the CAN decrypt algorithm will work for all the cards, as it's pretty complex and I only have 1 card (mine) to play with. But I haven't found any obvious parameters that could be card-dependent.

Magic_key is hardcoded in a DLL. So I assume it won't change any time soon?

In any case CAN can be stored & provided via the conf and/or env variable. What kind of parallel use did you have in mind?

Reply:
I see... With two million Slovenians and 10 million possible CANs, *knowing* a user's CAN allows tracking him around the country. I wrote about this extensively in [my masters thesis](http://sar.informatik.hu-berlin.de/research/publications/SAR-PR-2012-05/SAR-PR-2012-05_.pdf).

Anyway, if this is already present in the cards that are handed out, I agree that there will be no way of changing this any time soon.